### ISSUE TITLE
Nested assertions not reported

### ISSUE THREAD

**Issue by bobbicodes** [exercism/clojure-test-runner#22]
Issues have been filed like https://github.com/exercism/clojure/issues/420 and https://github.com/exercism/clojure/issues/421 which indicate that the test runner is not functioning properly. This was due to the limited scope of test suites it was tested against before release, and the richer use of nested assertions we're using in the Concept exercises.

There is still further implementation to be done to get it into a "complete" state, as it currently only handles individual `deftest` forms (see examples below) as was common in practice exercises. Forms containing nested assertions are tested properly, blocking the student if any assertions fail, but these nested assertions are not included in the results, causing mass confusion and hysteria.

One [cheesy] solution would be to modify the test suites to only contain individual `deftest`s. (see https://github.com/exercism/clojure/pull/422) This would not be ideal, but depending on how quickly the test runner can be upgraded it may be the quickest achievable path to a fully working track.

The work needing to be done is significant but doable. I intend to use this issue as an architectural log, to help gather my thoughts about how the implementation works and communicate them in a manner that can facilitate open collaboration.

**Comment by porkostomus** [comment]
The test runner works by using `clojure.test`, the de-facto Clojure testing framework. However, the amount of twisting it into shape necessary to get it to do what we need is pushing me to try a more direct approach.

It was no problem overriding the test reporting methods to output proper maps of the passes, fails and errors. However, they are reported out of order! This could be overcome by using a test fixture, but instead I opted to go straight to the source, and parse the test file and extract the tests in the order defined. Enter `rewrite-clj`. It worked spectacularly well, and made it trivial to also extract the assertion code to be provided in the output to the student.

My proposed solution to the current problem is to implement a recursive evaluator to traverse to test file's parse tree (a zipper data structure) and take over the job of `clojure.test` almost completely, merely using it to evaluate individual `is` assertions. We would then eliminate the custom reporting methods, because we would no longer be pushing it beyond the common use case.

**Comment by porkostomus** [comment]
The simplest Clojure test suite looks like this:

```clojure
(deftest year-not-divisible-by-4
  (is (not (leap/leap-year? 2015))))

(deftest year-divisible-by-2-but-not-4
  (is (not (leap/leap-year? 1970))))

(deftest year-divisible-by-4-but-not-100
  (is (leap/leap-year? 1996)))

(deftest year-divisible-by-4-and-5
  (is (leap/leap-year? 1960)))

(deftest year-divisible-by-100-but-not-400
  (is (not (leap/leap-year? 2100))))

(deftest year-divisible-by-100-but-not-by-3
  (is (not (leap/leap-year? 1900))))

(deftest year-divisible-by-400
  (is (leap/leap-year? 2000)))

(deftest year-divisible-by-400-but-not-125
  (is (leap/leap-year? 2400)))

(deftest year-divisible-by-200-but-not-by-400
  (is (not (leap/leap-year? 1800))))
```
This is the actual code that was used for the development of the current test runner, and that is the shape of the data that it expects. It parses the file, finds every `deftest`, and everybody's happy.

**Comment by porkostomus** [comment]
The problem arises when we get something like this:

```clojure
(deftest prep-time-test
  (testing "Preparation time in minutes"
    (testing "for one layer"
      (is (= 2 (lucians-luscious-lasagna/prep-time 1))))
    (testing "for multiple layers"
      (is (= 8 (lucians-luscious-lasagna/prep-time 4))))))
```
Here's where we need to recursively traverse the `deftest` node to grab each child node, going deeper if it's a `testing` until we hit an `is`, then we could say return a sequence of test strings and their evaluation results like this:

```clojure
[{:test-string "Preparation time in minutes for one layer"
  :result true}
 {:test-string "Preparation time in minutes for for multiple layers"
  :result true}]
```

**Comment by porkostomus** [comment]
Given an inner testing node, this function will find and evaluate its assertions:

```clojure
(defn results 
  "Traverses a zipper from an inner testing node
   and returns a sequence of its assertion results."
  [loc]
  (loop [loc loc
         assertions []]
    (cond
      (nil? loc) assertions
      (testing? loc) (recur (-> loc z/down z/right z/right) assertions)
      (assertion? loc) (recur (-> loc z/right) (conj assertions (assertion-true? loc))))))

;;=> [true true]
```

Now we just need to compose these and we're done.

**Comment by porkostomus** [comment]
The following evaluator seems to handle any `deftest` form:

```clojure
(def deftest-loc3
  (z/of-string
"(deftest signal-prisoner-test
  (testing \"Can signal prisoner if archer is sleeping and prisoner is awake\"
    (is (= true (annalyns-infiltration/can-signal-prisoner? false true))))
  (testing \"Cannot signal prisoner if\"
    (testing \"archer is awake and prisoner is sleeping\"
      (is (= false (annalyns-infiltration/can-signal-prisoner? true false))))
    (testing \"archer and prisoner are both sleeping\"
      (is (= false (annalyns-infiltration/can-signal-prisoner? false false))))
    (testing \"archer and prisoner are both awake\"
      (is (= false (annalyns-infiltration/can-signal-prisoner? true true))))))"))

(let [deftest deftest-loc3]
  (loop [loc deftest
         prefix-string ""
         test-strings []
         results []]
    (cond
      (deftest? loc)
      (recur (-> loc z/down z/right z/right)
             prefix-string test-strings results)
      (outer-testing? loc)
      (recur (-> loc z/down z/right z/right)
             (-> loc z/down z/right z/sexpr)
             test-strings results)
      (testing? loc)
      (recur (-> loc z/down z/right z/right)
             prefix-string
             (conj test-strings (str/trim (str prefix-string " " (-> loc z/down z/right z/sexpr))))
             (conj results []))
      (and (assertion? loc) (assertion-next? loc))
      (recur (-> loc z/right)
             prefix-string
             test-strings
             (conj results [(assertion-true? loc)]))
      (assertion? loc)
      (recur (-> loc z/up z/right)
             prefix-string
             test-strings
             (conj (vec (butlast results))
                   (conj (vec (last results)) (assertion-true? loc))))
      :else
      {:test-name (-> deftest z/down z/right z/sexpr str)
       :results (vec (remove empty? results))
       :test-strings test-strings})))

;;=> {:test-name "signal-prisoner-test",
 :results [[true] [true] [true] [true]],
 :test-strings
 ["Can signal prisoner if archer is sleeping and prisoner is awake"
  "Cannot signal prisoner if archer is awake and prisoner is sleeping"
  "Cannot signal prisoner if archer and prisoner are both sleeping"
  "Cannot signal prisoner if archer and prisoner are both awake"]}
```

**Comment by ErikSchierboom** [comment]
👍

**Comment by RoelofWobben** [comment]
Hope that I quick can test things. 
WIth the pr that made all test one by one , I still cannot see why things are failing

**Comment by miridius** [comment]
Hey so I'm not sure if we really need to reinvent the wheel here and recreate clojure.test. In #42 I've updated the existing solution to support multiple `is` forms inside a `deftest`, although it currently assumes every top level form is an assertion I think it would be pretty easy to update it to handle nesting.

If we do decide that clojure.test is too restrictive then we could always look for a different off the shelf test runner instead of writing a new one!

**Comment by bobbicodes** [comment]
Fixed by #42
